### ISSUE TITLE
Test failures in the `synthetix` repo

### ISSUE THREAD

**Issue by agostbiro** [NomicFoundation/edr#207]
Multiple tests fail in the [synthetix](https://github.com/Synthetixio/synthetix) repo when ran against the `rethnet/main` branch.

**Steps to reproduce**

1. Check out  [synthetix](https://github.com/Synthetixio/synthetix) @ `3040126f`
2. Link `synthetix` to `rethnet/main` version of Hardhat (tested with `bd196ec07 `)
3. Run `npx hardhat test ` in the synthetix repo

**Expected result**

The tests pass.

**Actual result**

Multiple tests fail: https://pastebin.com/v0R1knaa

**Notes**

Tests pass when ran again `main`, except for a sporadic failure which was observed in one in three runs against `main`:

```
 1) Contract: PerpsV2Market PerpsV2MarketAtomic
       PerpsV2 Liquidations
         liquidatePosition
           liquidations of positive margin position pays to fee pool, long case:

      AssertionError: expected '227400150000006362750' to equal '227400150000003014000'
      + expected - actual

      -227400150000006362750
      +227400150000003014000

      at Object.assertBNEqual [as bnEqual] (test/utils/index.js:348:10)
      at Context.<anonymous> (test/contracts/PerpsV2Market.js:5796:12)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)

  2) Contract: PerpsV2Market PerpsV2MarketAtomic
       PerpsV2 Liquidations
         liquidatePosition
           liquidations of positive margin position pays to fee pool, long case:

      AssertionError: expected '227400150000007367500' to equal '227400150000003014000'
      + expected - actual

      -227400150000007367500
      +227400150000003014000

      at Object.assertBNEqual [as bnEqual] (test/utils/index.js:348:10)
      at Context.<anonymous> (test/contracts/PerpsV2Market.js:5850:12)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by Wodann** [comment]
I tested this with the fix from https://github.com/NomicFoundation/hardhat/pull/4499. In a single test run, the only failure I had is similar to the failure you saw on `main`:

```
1) Contract: PerpsV2Market PerpsV2MarketAtomic
       PerpsV2 Liquidations
         liquidatePosition
           liquidations of positive margin position pays to fee pool, long case:

      AssertionError: expected '227400150000006362750' to equal '227400150000003014000'
      + expected - actual

      -227400150000006362750
      +227400150000003014000
      
      at Object.assertBNEqual [as bnEqual] (test/utils/index.js:348:10)
      at Context.<anonymous> (test/contracts/PerpsV2Market.js:5796:12)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by agostbiro** [comment]
I tested again on the benchmark server with https://github.com/NomicFoundation/hardhat/pull/4499 and I can still repro the errors reported in the [pastebin.](https://pastebin.com/v0R1knaa)

Quicker command to repro: `npx hardhat test --grep "should set constructor params on deployment"`

Erros with

```
     HardhatError: HH11: An internal invariant was violated: Event beforeMessage is not supported in node._vm
      at assertHardhatInvariant (/home/ubuntu/dev/hardhat/packages/hardhat-core/src/internal/core/errors.ts:272:11)
      at Proxy.on (/home/ubuntu/dev/hardhat/packages/hardhat-core/src/internal/hardhat-network/provider/vm/proxy-vm.ts:87:33)
      at Function.fromEvent (/home/ubuntu/dev/synthetix/node_modules/@defi-wonderland/smock/src/observable-vm.ts:35:20)
      at new ObservableVM (/home/ubuntu/dev/synthetix/node_modules/@defi-wonderland/smock/src/observable-vm.ts:17:40)
      at new Sandbox (/home/ubuntu/dev/synthetix/node_modules/@defi-wonderland/smock/src/sandbox.ts:36:15)
      at Function.create (/home/ubuntu/dev/synthetix/node_modules/@defi-wonderland/smock/src/sandbox.ts:82:12)
      at init (/home/ubuntu/dev/synthetix/node_modules/@defi-wonderland/smock/src/index.ts:27:13)
      at Object.fake (/home/ubuntu/dev/synthetix/node_modules/@defi-wonderland/smock/src/index.ts:13:17)
      at Context.<anonymous> (/home/ubuntu/dev/synthetix/test/contracts/SynthetixBridgeToOptimism.unit.js:60:16)
```

I think it's due to https://github.com/NomicFoundation/hardhat/pull/4481

**Comment by Wodann** [comment]
This should have been fixed by my JS callback PR

**Comment by agostbiro** [comment]
From my latest testing: 

The only error I observed in the Synthetix repo when running the test suite with `rethnet/main` is this:

```
  1) Contract: PerpsV2Market PerpsV2MarketAtomic
       PerpsV2 Liquidations
         liquidatePosition
           liquidations of positive margin position pays to fee pool, long case:

      AssertionError: expected '227400150000007367500' to equal '227400150000003014000'
      + expected - actual

      -227400150000007367500
      +227400150000003014000
```

I observed this twice in a row when running the entire Synthetix test suite against `rethnet/main@94e75d882` with `npx hardhat test`. I didn't observe this failure when running the entire Synthetix test suite twice with their Hardhat version `2.12.7`. I did however observe it one out of three times when running the Synthetix test suite against `main`.

Note that I couldn't reproduce the error with multiple runs of:

`npx hardhat test --grep "liquidations of positive margin position pays to fee pool, long case"` (runs only the specific unit test)

or

`npx hardhat test --grep "Contract: PerpsV2Market PerpsV2MarketAtomic"` (runs the entire contract test suite in which the sporadic unit test failure was observed)
